File: Python3/0519_random_flip_matrix.py
# https://leetcode.cn/problems/random-flip-matrix/

class Solution:

    def __init__(self, m: int, n: int):
        self.m = m
        self.n = n
        self.last_idx = m * n - 1
        self.zeros = m * n
        self.ones = set()

    def flip(self) -> List[int]:
        idx = random.randint(0, self.zeros-1)
        if idx == self.last_idx or idx in self.ones:
            while self.last_idx in self.ones:
                self.last_idx -= 1
            res = [self.last_idx // self.n, self.last_idx % self.n]
            self.last_idx -= 1            
        else:
            self.ones.add(idx)
            res = [idx // self.n, idx % self.n]
        self.zeros -= 1
        return res

    def reset(self) -> None:
        self.last_idx = self.m * self.n - 1
        self.zeros = self.m * self.n
        self.ones.clear()


# Your Solution object will be instantiated and called as such:
# obj = Solution(m, n)
# param_1 = obj.flip()
# obj.reset()

File: Cpp/0519_random_flip_matrix.cpp
// https://leetcode.cn/problems/random-flip-matrix/

class Solution {
public:
    Solution(int m, int n) {
        m_ = m;
        n_ = n;
        zeros_ = m * n;
        last_idx_ = zeros_ - 1;
    }
    
    vector<int> flip() {
        int idx = rand() % zeros_;
        vector<int> res(2);
        if (idx == last_idx_ || ones_.find(idx) != ones_.end()) {
            while (ones_.find(last_idx_) != ones_.end()) {
                last_idx_ -= 1;
            }
            res[0] = last_idx_ / n_;
            res[1] = last_idx_ % n_;
            last_idx_ -= 1;
        }
        else {
            ones_.insert(idx);
            res[0] = idx / n_;
            res[1] = idx % n_;
        }
        zeros_ -= 1;
        return res;
    }
    
    void reset() {
        zeros_ = m_ * n_;
        last_idx_ = zeros_ - 1;
        ones_.clear();
    }

    int m_, n_, zeros_;
    int last_idx_;
    unordered_set<int> ones_;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(m, n);
 * vector<int> param_1 = obj->flip();
 * obj->reset();
 */
